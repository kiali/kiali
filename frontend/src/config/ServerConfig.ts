import _ from 'lodash';
import { ServerConfig } from '../types/ServerConfig';
import { parseHealthConfig } from './HealthConfig';
import { MeshCluster } from '../types/Mesh';

export type Durations = { [key: number]: string };

export type ComputedServerConfig = ServerConfig & {
  durations: Durations;
};

function getHomeCluster(cfg: ServerConfig): MeshCluster | undefined {
  return Object.values(cfg.clusters).find(cluster => cluster.isKialiHome);
}

export const humanDurations = (cfg: ComputedServerConfig, prefix?: string, suffix?: string): Durations =>
  _.mapValues(cfg.durations, v => _.reject([prefix, v, suffix], _.isEmpty).join(' '));

const toDurations = (tupleArray: [number, string][]): Durations => {
  const obj: Duration = {};
  tupleArray.forEach(tuple => {
    obj[tuple[0]] = tuple[1];
  });
  return obj;
};

const durationsTuples: [number, string][] = [
  [60, '1m'],
  [120, '2m'],
  [300, '5m'],
  [600, '10m'],
  [1800, '30m'],
  [3600, '1h'],
  [10800, '3h'],
  [21600, '6h'],
  [43200, '12h'],
  [86400, '1d'],
  [604800, '7d'],
  [2592000, '30d']
];

const computeValidDurations = (cfg: ComputedServerConfig): void => {
  const tsdbRetention = cfg.prometheus.storageTsdbRetention;
  const scrapeInterval = cfg.prometheus.globalScrapeInterval;
  let filtered = durationsTuples.filter(
    d => (!tsdbRetention || d[0] <= tsdbRetention!) && (!scrapeInterval || d[0] >= scrapeInterval * 2)
  );
  // Make sure we keep at least one item, even if it's silly
  if (filtered.length === 0) {
    filtered = [durationsTuples[0]];
  }
  cfg.durations = toDurations(filtered);
};

// Set some reasonable defaults. Initial values should be valid for fields
// than may not be providedby/set on the server.
const defaultServerConfig: ComputedServerConfig = {
  accessibleNamespaces: [],
  ambientEnabled: false,
  authStrategy: '',
  clusters: {},
  durations: {},
  gatewayAPIClasses: [],
  gatewayAPIEnabled: false,
  logLevel: '',
  healthConfig: {
    rate: []
  },
  deployment: {
    viewOnlyMode: false
  },
  installationTag: 'Kiali Console',
  istioAnnotations: {
    ambientAnnotation: 'ambient.istio.io/redirection',
    ambientAnnotationEnabled: 'enabled',
    istioInjectionAnnotation: 'sidecar.istio.io/inject'
  },
  istioCanaryRevision: {
    current: '',
    upgrade: ''
  },
  istioIdentityDomain: 'svc.cluster.local',
  istioNamespace: 'istio-system',
  istioLabels: {
    ambientNamespaceLabel: 'istio.io/dataplane-mode',
    ambientNamespaceLabelValue: 'ambient',
    ambientWaypointLabel: 'gateway.istio.io/managed',
    ambientWaypointLabelValue: 'istio.io-mesh-controller',
    appLabelName: 'app',
    egressGatewayLabel: 'istio=egressgateway',
    ingressGatewayLabel: 'istio=ingressgateway',
    injectionLabelName: 'istio-injection',
    injectionLabelRev: 'istio.io/rev',
    versionLabelName: 'version'
  },
  kialiFeatureFlags: {
    certificatesInformationIndicators: {
      enabled: true
    },
    disabledFeatures: [],
    istioInjectionAction: true,
    istioAnnotationAction: true,
    istioUpgradeAction: false,
    uiDefaults: {
      graph: {
        findOptions: [],
        hideOptions: [],
        impl: 'cy',
        settings: {
          fontLabel: 13,
          minFontBadge: 7,
          minFontLabel: 10
        },
        traffic: {
          grpc: 'requests',
          http: 'requests',
          tcp: 'sent'
        }
      },
      i18n: {
        language: 'en',
        showSelector: false
      },
      list: {
        includeHealth: true,
        includeIstioResources: true,
        includeValidations: true,
        showIncludeToggles: false
      },
      mesh: {
        findOptions: [],
        hideOptions: [],
        impl: 'classic'
      }
    }
  },
  prometheus: {
    globalScrapeInterval: 15,
    storageTsdbRetention: 21600
  }
};

// Overwritten with real server config on user login. Also used for tests.
let serverConfig = defaultServerConfig;
computeValidDurations(serverConfig);
export { serverConfig };

let homeCluster = getHomeCluster(serverConfig);
let isMultiCluster = isMC();
export { homeCluster, isMultiCluster };

export const toValidDuration = (duration: number): number => {
  // Check if valid
  if (serverConfig.durations[duration]) {
    return duration;
  }
  // Get closest duration
  const validDurations = durationsTuples.filter(d => serverConfig.durations[d[0]]);
  for (let i = validDurations.length - 1; i > 0; i--) {
    if (duration > durationsTuples[i][0]) {
      return validDurations[i][0];
    }
  }
  return validDurations[0][0];
};

export const setServerConfig = (cfg: ServerConfig): void => {
  serverConfig = {
    ...defaultServerConfig,
    ...cfg
  };

  serverConfig.healthConfig = cfg.healthConfig ? parseHealthConfig(cfg.healthConfig) : serverConfig.healthConfig;
  computeValidDurations(serverConfig);

  homeCluster = getHomeCluster(serverConfig);
  isMultiCluster = isMC();
};

export const isIstioNamespace = (namespace: string): boolean => {
  if (namespace === serverConfig.istioNamespace) {
    return true;
  }
  return false;
};

export const isHomeCluster = (cluster: string): boolean => {
  return !isMultiCluster || cluster === homeCluster?.name;
};

// Return true if the cluster is configured for this Kiali instance
export const isConfiguredCluster = (cluster: string): boolean => {
  return Object.keys(serverConfig.clusters).includes(cluster);
};

function isMC(): boolean {
  // If there is only one cluster, it is not a multi-cluster deployment.
  // If there are multiple clusters but only one is accessible, it is not a multi-cluster deployment.
  return (
    Object.keys(serverConfig.clusters).length > 1 &&
    Object.values(serverConfig.clusters).filter(c => c.accessible).length > 1
  );
}
