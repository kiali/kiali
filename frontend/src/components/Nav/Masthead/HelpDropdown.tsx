import * as React from 'react';
import { AboutUIModal } from '../../About/AboutUIModal';
import { KialiAppState } from '../../../store/Store';
import { DebugInformation } from '../../../components/DebugInformation/DebugInformation';
import { QuestionCircleIcon } from '@patternfly/react-icons/';
import { connect } from 'react-redux';
import { isUpstream } from '../../UpstreamDetector/UpstreamDetector';
import { Status, StatusKey } from '../../../types/StatusState';
import { config, serverConfig } from '../../../config';
import { IstioCertsInfo } from 'components/IstioCertsInfo/IstioCertsInfo';
import { Dropdown, DropdownItem, DropdownList, MenuToggle, MenuToggleElement } from '@patternfly/react-core';
import { useKialiTranslation } from 'utils/I18nUtils';

type ReduxProps = {
  status: Status;
  warningMessages: string[];
};

type HelpDropdownProps = ReduxProps;

const HelpDropdownComponent: React.FC<HelpDropdownProps> = (props: HelpDropdownProps) => {
  const [isDropdownOpen, setIsDropdownOpen] = React.useState<boolean>(false);
  const [isAboutModalOpen, setIsAboutModalOpen] = React.useState<boolean>(false);
  const [isDebugInformationOpen, setIsDebugInformationOpen] = React.useState<boolean>(false);
  const [isCertsInformationOpen, setIsCertsInformationOpen] = React.useState<boolean>(false);

  const { t } = useKialiTranslation();

  const onDropdownSelect = (): void => {
    setIsDropdownOpen(!isDropdownOpen);
  };

  const buildDocumentationLink = (): string => {
    const url = new URL(config.documentation.url);

    if (isUpstream) {
      const kialiCoreVersion = props.status[StatusKey.KIALI_CORE_VERSION] || 'unknown';

      url.searchParams.append('utm_source', 'kiali');
      url.searchParams.append('utm_medium', 'app');
      url.searchParams.append('utm_campaign', kialiCoreVersion);
      url.searchParams.append('utm_content', '?-menu');
    }

    return url.toString();
  };

  const items: React.ReactNode[] = [];

  items.push(
    <DropdownItem key={'view_documentation'} onClick={() => window.open(buildDocumentationLink(), '_blank')}>
      Documentation
    </DropdownItem>
  );

  items.push(
    <DropdownItem key={'view_debug_info'} onClick={() => setIsDebugInformationOpen(true)}>
      View Debug Info
    </DropdownItem>
  );

  if (serverConfig.kialiFeatureFlags.certificatesInformationIndicators.enabled) {
    items.push(
      <DropdownItem key={'view_certs_info'} onClick={() => setIsCertsInformationOpen(true)}>
        View Certificates Info
      </DropdownItem>
    );
  }

  items.push(
    <DropdownItem key={'view_about_info'} onClick={() => setIsAboutModalOpen(true)}>
      About
    </DropdownItem>
  );

  return (
    <>
      <AboutUIModal
        status={props.status}
        warningMessages={props.warningMessages}
        isOpen={isAboutModalOpen}
        onClose={() => setIsAboutModalOpen(false)}
      />

      <DebugInformation isOpen={isDebugInformationOpen} onClose={() => setIsDebugInformationOpen(false)} />

      {serverConfig.kialiFeatureFlags.certificatesInformationIndicators.enabled && (
        <IstioCertsInfo isOpen={isCertsInformationOpen} onClose={() => setIsCertsInformationOpen(false)} />
      )}

      <Dropdown
        toggle={(toggleRef: React.Ref<MenuToggleElement>) => (
          <MenuToggle
            ref={toggleRef}
            data-test="about-help-button"
            onClick={() => setIsDropdownOpen(!isDropdownOpen)}
            aria-label={t('Help')}
            variant="plain"
            isExpanded={isDropdownOpen}
          >
            <QuestionCircleIcon />
          </MenuToggle>
        )}
        isOpen={isDropdownOpen}
        popperProps={{ position: 'right' }}
        onOpenChange={(isOpen: boolean) => setIsDropdownOpen(isOpen)}
        onSelect={onDropdownSelect}
      >
        <DropdownList>{items}</DropdownList>
      </Dropdown>
    </>
  );
};

const mapStateToProps = (state: KialiAppState): ReduxProps => ({
  status: state.statusState.status,
  warningMessages: state.statusState.warningMessages
});

export const HelpDropdown = connect(mapStateToProps)(HelpDropdownComponent);
