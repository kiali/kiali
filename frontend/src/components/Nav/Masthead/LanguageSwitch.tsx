import * as React from 'react';
import { kialiStyle } from 'styles/StyleUtils';
import { Dropdown, DropdownItem, DropdownList, MenuToggle, MenuToggleElement, Tooltip } from '@patternfly/react-core';
import { KialiIcon } from 'config/KialiIcon';
import { GlobalActions } from 'actions/GlobalActions';
import { store } from 'store/ConfigStore';
import { Language } from 'types/Common';
import { KialiAppState } from 'store/Store';
import { connect } from 'react-redux';
import { i18n } from 'i18n';
import { useKialiTranslation } from 'utils/I18nUtils';

const iconStyle = kialiStyle({
  verticalAlign: '-0.25rem',
  $nest: {
    '& svg': {
      width: '1.5rem',
      height: '1.5rem'
    }
  }
});

const checkStyle = kialiStyle({
  marginLeft: '0.5rem'
});

type LanguageSwitchProps = {
  language: string;
};

export const LanguageSwitchComponent: React.FC<LanguageSwitchProps> = ({ language }) => {
  const [isDropdownOpen, setIsDropdownOpen] = React.useState<boolean>(false);

  const { t } = useKialiTranslation();

  const languageItems: React.ReactNode[] = [
    <DropdownItem key="English" onClick={() => switchLanguage(Language.ENGLISH)}>
      <span>English</span>
      {language === Language.ENGLISH && <KialiIcon.Check className={checkStyle} />}
    </DropdownItem>,

    <DropdownItem key="Chinese" onClick={() => switchLanguage(Language.CHINESE)}>
      <span>中文</span>
      {language === Language.CHINESE && <KialiIcon.Check className={checkStyle} />}
    </DropdownItem>
  ];

  const switchLanguage = (language: string): void => {
    i18n.changeLanguage(language).then(() => store.dispatch(GlobalActions.setLanguage(language)));
  };

  const onDropdownSelect = (): void => {
    setIsDropdownOpen(!isDropdownOpen);
  };

  return (
    <Tooltip position="bottom" content={<>{t('Switch language')}</>} trigger="mouseenter click" exitDelay={0}>
      <Dropdown
        toggle={(toggleRef: React.Ref<MenuToggleElement>) => (
          <MenuToggle
            ref={toggleRef}
            data-test="switch-language-button"
            onClick={() => setIsDropdownOpen(!isDropdownOpen)}
            aria-label={t('Switch language')}
            variant="plain"
            isExpanded={isDropdownOpen}
          >
            <KialiIcon.Language className={iconStyle} />
          </MenuToggle>
        )}
        isOpen={isDropdownOpen}
        popperProps={{ position: 'center' }}
        onOpenChange={(isOpen: boolean) => setIsDropdownOpen(isOpen)}
        onSelect={onDropdownSelect}
      >
        <DropdownList>{languageItems}</DropdownList>
      </Dropdown>
    </Tooltip>
  );
};

const mapStateToProps = (state: KialiAppState): LanguageSwitchProps => {
  return {
    language: state.globalState.language
  };
};

export const LanguageSwitch = connect(mapStateToProps)(LanguageSwitchComponent);
