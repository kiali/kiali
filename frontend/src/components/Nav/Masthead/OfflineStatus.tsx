import * as React from 'react';
import { DisconnectedIcon } from '@patternfly/react-icons';
import { Tooltip, Label } from '@patternfly/react-core';
import { serverConfig } from 'config';
import { kialiStyle } from 'styles/StyleUtils';
import { PFColors } from 'components/Pf/PfColors';

const labelTextStyle = kialiStyle({
  color: PFColors.Blue400
});

export const OfflineStatus: React.FC = () => {
  const date = new Date(serverConfig.runConfig!.timestamp!);
  let content = `You are running Kiali offline. Some features may be degraded or unavailable in this mode. This is a snapshot of your environment.`;
  if (serverConfig.runConfig?.timestamp !== undefined) {
    // Trim '.' to add snapshot time/date.
    content = content.substring(0, content.length - 1);
    content += ` taken at ${date.toLocaleTimeString()} on ${date.toLocaleDateString()}.`;
  }
  return (
    <Tooltip content={content}>
      <Label color="grey" icon={<DisconnectedIcon color={PFColors.Blue400} data-test="offline-status" />}>
        <span className={labelTextStyle}>offline</span>
      </Label>
    </Tooltip>
  );
};
