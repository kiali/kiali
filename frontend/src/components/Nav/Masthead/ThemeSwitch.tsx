import * as React from 'react';
import { Button, Tooltip } from '@patternfly/react-core';
import { KialiAppState } from 'store/Store';
import { connect } from 'react-redux';
import { PF_THEME_DARK, Theme } from 'types/Common';
import { GlobalActions } from 'actions/GlobalActions';
import { store } from 'store/ConfigStore';
import { kialiStyle } from 'styles/StyleUtils';
import { PFColors } from 'components/Pf/PfColors';
import { KialiIcon } from 'config/KialiIcon';
import { classes } from 'typestyle';
import { useKialiTranslation } from 'utils/I18nUtils';

const iconStyle = kialiStyle({
  color: PFColors.Color100
});

const buttonStyle = kialiStyle({
  fontSize: '1rem',
  $nest: {
    '&.pf-m-primary': {
      backgroundColor: PFColors.Blue400,
      $nest: {
        '&::after': {
          border: `1px solid ${PFColors.Blue200}`
        }
      }
    },
    '&.pf-m-secondary::after': {
      border: `1px solid ${PFColors.Color400}`
    }
  }
});

const lightButtonStyle = kialiStyle({
  borderTopRightRadius: 0,
  borderBottomRightRadius: 0,
  $nest: {
    '&::after': {
      borderTopRightRadius: 0,
      borderBottomRightRadius: 0
    },
    '&&.pf-m-secondary::after': {
      borderRight: 0
    }
  }
});

const darkButtonStyle = kialiStyle({
  borderTopLeftRadius: 0,
  borderBottomLeftRadius: 0,
  $nest: {
    '&::after': {
      borderTopLeftRadius: 0,
      borderBottomLeftRadius: 0
    },
    '&&.pf-m-secondary::after': {
      borderLeft: 0
    }
  }
});

type ThemeSwitchProps = {
  theme: string;
};

export const ThemeSwitchComponent: React.FC<ThemeSwitchProps> = (props: ThemeSwitchProps) => {
  const { t } = useKialiTranslation();

  const darkTheme = props.theme === Theme.DARK;

  const handleTheme = (): void => {
    const theme = darkTheme ? Theme.LIGHT : Theme.DARK;

    document.documentElement.classList.toggle(PF_THEME_DARK);
    store.dispatch(GlobalActions.setTheme(theme));
  };

  return (
    <Tooltip position="bottom" content={<>{t('Switch to {{theme}} Mode', { theme: darkTheme ? 'Light' : 'Dark' })}</>}>
      <div>
        <Button
          variant={darkTheme ? 'secondary' : 'primary'}
          className={classes(buttonStyle, lightButtonStyle)}
          onClick={handleTheme}
        >
          <KialiIcon.Sun className={iconStyle}></KialiIcon.Sun>
        </Button>
        <Button
          variant={darkTheme ? 'primary' : 'secondary'}
          className={classes(buttonStyle, darkButtonStyle)}
          onClick={handleTheme}
        >
          <KialiIcon.Moon className={iconStyle}></KialiIcon.Moon>
        </Button>
      </div>
    </Tooltip>
  );
};

const mapStateToProps = (state: KialiAppState): ThemeSwitchProps => {
  return {
    theme: state.globalState.theme
  };
};

export const ThemeSwitch = connect(mapStateToProps)(ThemeSwitchComponent);
