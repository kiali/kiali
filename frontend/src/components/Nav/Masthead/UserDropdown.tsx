import * as React from 'react';
import { SessionTimeout } from '../../SessionTimeout/SessionTimeout';
import { config } from '../../../config';
import { isMultiCluster } from '../../../config';
import { MILLISECONDS } from '../../../types/Common';
import { Timer } from 'globals';
import { KialiAppState, LoginSession } from '../../../store/Store';
import { authenticationConfig } from '../../../config/AuthenticationConfig';
import { AuthStrategy } from '../../../types/Auth';
import moment from 'moment';
import { KialiDispatch } from 'types/Redux';
import { LoginThunkActions } from '../../../actions/LoginThunkActions';
import { connect } from 'react-redux';
import * as API from '../../../services/Api';
import { kialiStyle } from 'styles/StyleUtils';
import { namespacesPerClusterSelector } from 'store/Selectors';
import {
  Divider,
  Dropdown,
  DropdownGroup,
  DropdownItem,
  DropdownList,
  MenuToggle,
  MenuToggleElement
} from '@patternfly/react-core';
import { t } from 'utils/I18nUtils';

type ReduxStateProps = {
  clusters?: string[];
  session?: LoginSession;
};

type ReduxDispatchProps = {
  extendSession: (session: LoginSession) => void;
  logout: () => void;
};

type UserProps = ReduxStateProps & ReduxDispatchProps;

type UserState = {
  checkSessionTimerId?: Timer;
  isDropdownOpen: boolean;
  isSessionTimeoutDismissed: boolean;
  showSessionTimeOut: boolean;
  timeCountDownSeconds: number;
  timeLeftTimerId?: Timer;
};

const dropdownStyle = kialiStyle({
  paddingLeft: 0,
  paddingRight: 0
});

class UserDropdownComponent extends React.Component<UserProps, UserState> {
  constructor(props: UserProps) {
    super(props);

    this.state = {
      showSessionTimeOut: false,
      timeCountDownSeconds: this.timeLeft() / MILLISECONDS,
      isSessionTimeoutDismissed: false,
      isDropdownOpen: false
    };
  }

  componentDidMount(): void {
    const checkSessionTimerId = setInterval(() => {
      this.checkSession();
    }, 3000);

    const timeLeftTimerId = setInterval(() => {
      this.setState({ timeCountDownSeconds: this.timeLeft() / MILLISECONDS });
    }, 1000);

    this.setState({
      checkSessionTimerId: checkSessionTimerId,
      timeLeftTimerId: timeLeftTimerId
    });
  }

  componentWillUnmount(): void {
    if (this.state.checkSessionTimerId) {
      clearInterval(this.state.checkSessionTimerId);
    }

    if (this.state.timeLeftTimerId) {
      clearInterval(this.state.timeLeftTimerId);
    }
  }

  timeLeft = (): number => {
    const expiresOn = moment(this.props.session!.expiresOn);

    if (expiresOn <= moment()) {
      this.props.logout();
    }

    return expiresOn.diff(moment());
  };

  checkSession = (): void => {
    if (this.timeLeft() < config.session.timeOutforWarningUser) {
      this.setState({ showSessionTimeOut: true });
    }
  };

  handleLogout = (): void => {
    if (authenticationConfig.logoutEndpoint) {
      API.logout()
        .then(_ => {
          (document.getElementById('openshiftlogout') as HTMLFormElement).submit();
        })
        .catch(error => {
          const errorMsg = error.response && error.response.data.error ? error.response.data.error : error.message;
          console.error(`Logout failed. "kiali-token" cookie may need to be cleared manually: ${errorMsg}`);
          (document.getElementById('openshiftlogout') as HTMLFormElement).submit();
        });
    } else {
      this.props.logout();
    }
  };

  extendSession = (session: LoginSession): void => {
    this.props.extendSession(session);
    this.setState({ showSessionTimeOut: false });
  };

  onDropdownToggle = (isDropdownOpen: boolean): void => {
    this.setState({
      isDropdownOpen
    });
  };

  onDropdownSelect = (): void => {
    this.setState({
      isDropdownOpen: !this.state.isDropdownOpen
    });
  };

  render(): React.ReactNode {
    const { isDropdownOpen } = this.state;

    const clusterIsInSessionInfo = (cluster: string): boolean =>
      this.props.session?.clusterInfo?.[cluster] !== undefined;

    const canLogout =
      authenticationConfig.strategy !== AuthStrategy.anonymous && authenticationConfig.strategy !== AuthStrategy.header;

    // We want to show a dropdown per cluster the user is not logged into yet.
    // The clusters you are logged into are in session.clusterInfo and all
    // the clusters are in authenticationConfig.authorizationEndpointPerCluster.
    // So the clusters you are not logged into is authenticationConfig.authorizationEndpointPerCluster - session.clusterInfo.
    // Two groups of clusters: those you are logged into and those you are not.
    let loggedInClusters: { cluster: string; endpoint: string }[] = [];
    let loggedOutClusters: { cluster: string; endpoint: string }[] = [];
    if (authenticationConfig.authorizationEndpointPerCluster !== undefined) {
      Object.entries(authenticationConfig.authorizationEndpointPerCluster).forEach(([cluster, endpoint]) => {
        clusterIsInSessionInfo(cluster)
          ? loggedInClusters.push({ cluster: cluster, endpoint: endpoint })
          : loggedOutClusters.push({ cluster: cluster, endpoint: endpoint });
      });
    }

    return (
      <>
        <SessionTimeout
          onLogout={this.props.logout}
          onExtendSession={this.extendSession}
          onDismiss={this.dismissHandler}
          show={this.state.showSessionTimeOut && !this.state.isSessionTimeoutDismissed}
          timeOutCountDown={this.state.timeCountDownSeconds}
        />

        {this.props.session && !canLogout && <>{this.props.session.username}</>}

        {this.props.session && canLogout && (
          <Dropdown
            toggle={(toggleRef: React.Ref<MenuToggleElement>) => (
              <MenuToggle
                ref={toggleRef}
                className={dropdownStyle}
                variant="plainText"
                onClick={() => this.onDropdownToggle(!isDropdownOpen)}
                isExpanded={isDropdownOpen}
              >
                {this.props.session?.username}
              </MenuToggle>
            )}
            isOpen={isDropdownOpen}
            onSelect={this.onDropdownSelect}
            popperProps={{ position: 'right' }}
            onOpenChange={(isOpen: boolean) => this.onDropdownToggle(isOpen)}
          >
            {isMultiCluster && loggedInClusters.length > 0 && (
              <>
                <DropdownGroup label="logged-in clusters" labelHeadingLevel="h3">
                  <DropdownList>
                    {loggedInClusters.map(clusterInfo => {
                      return (
                        <DropdownItem isDisabled={true} key={clusterInfo.cluster}>
                          {clusterInfo.cluster}
                        </DropdownItem>
                      );
                    })}
                  </DropdownList>
                </DropdownGroup>
                <Divider component="li" />
              </>
            )}

            {isMultiCluster && loggedOutClusters.length > 0 && (
              <>
                <DropdownGroup label="logged-out clusters" labelHeadingLevel="h3">
                  <DropdownList>
                    {loggedOutClusters.map(clusterInfo => {
                      return (
                        <DropdownItem key={clusterInfo.cluster} to={clusterInfo.endpoint}>
                          {t('Login to {{cluster}}', { cluster: clusterInfo.cluster })}
                        </DropdownItem>
                      );
                    })}
                  </DropdownList>
                </DropdownGroup>
                <Divider component="li" />
              </>
            )}

            <DropdownItem key={'user_logout_option'} onClick={this.handleLogout} isDisabled={!canLogout}>
              {t('Logout')}
            </DropdownItem>
          </Dropdown>
        )}

        {authenticationConfig.strategy === AuthStrategy.openshift && authenticationConfig.logoutEndpoint && (
          <form id="openshiftlogout" action={authenticationConfig.logoutEndpoint} method="post">
            <input type="hidden" name="then" value={authenticationConfig.logoutRedirect} />
          </form>
        )}
      </>
    );
  }

  private dismissHandler = (): void => {
    this.setState({ isSessionTimeoutDismissed: true });
  };
}

const mapStateToProps = (state: KialiAppState): ReduxStateProps => ({
  clusters: Array.from(namespacesPerClusterSelector(state).keys()),
  session: state.authentication.session
});

const mapDispatchToProps = (dispatch: KialiDispatch): ReduxDispatchProps => ({
  logout: () => dispatch(LoginThunkActions.logout()),
  extendSession: (session: LoginSession) => dispatch(LoginThunkActions.extendSession(session))
});

export const UserDropdown = connect(mapStateToProps, mapDispatchToProps)(UserDropdownComponent);
