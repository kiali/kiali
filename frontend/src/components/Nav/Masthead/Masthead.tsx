import * as React from 'react';
import { Label, Flex, FlexItem, Tooltip, Toolbar, ToolbarItem } from '@patternfly/react-core';

import { homeCluster, serverConfig } from '../../../config';
import { IstioStatus } from '../../IstioStatus/IstioStatus';
import { UserDropdown } from './UserDropdown';
import { HelpDropdown } from './HelpDropdown';
import { MessageCenterTrigger } from '../../../components/MessageCenter/MessageCenterTrigger';
import { ThemeSwitch } from './ThemeSwitch';
import { LanguageSwitch } from './LanguageSwitch';
import { KialiIcon } from 'config/KialiIcon';
import { useKialiTranslation } from 'utils/I18nUtils';
import { PfSpinner } from 'components/Pf/PfSpinner';
import { kialiStyle } from 'styles/StyleUtils';

export const MASTHEAD = 'masthead';

const toolbarStyle = kialiStyle({
  marginLeft: 'auto',
  $nest: {
    '& .pf-v5-svg': {
      fontSize: '1rem'
    }
  }
});

const themeSwitchStyle = kialiStyle({
  marginLeft: '1.5rem',
  marginRight: 0
});

const messageCenterStyle = kialiStyle({
  marginRight: '0.25rem'
});

const helpDropdownStyle = kialiStyle({
  marginRight: '0.5rem'
});

const languageSwitchStyle = kialiStyle({
  marginRight: '0.75rem'
});

const userDropdownStyle = kialiStyle({
  marginLeft: '0.5rem',
  position: 'relative',
  bottom: '0.125rem'
});

export const MastheadItems: React.FC = () => {
  const { t } = useKialiTranslation();

  return (
    <>
      <PfSpinner />
      <Toolbar>
        <ToolbarItem className={toolbarStyle}>
          <Flex>
            <FlexItem align={{ default: 'alignRight' }}>
              {homeCluster?.name && (
                <Tooltip
                  entryDelay={0}
                  position="bottom"
                  content={<div>{t('Kiali home cluster: {{name}}', { name: homeCluster?.name })}</div>}
                >
                  <Label data-test="cluster-icon" color="blue" icon={<KialiIcon.Cluster />}>
                    {homeCluster?.name}
                  </Label>
                </Tooltip>
              )}
            </FlexItem>

            <FlexItem className={themeSwitchStyle}>
              <ThemeSwitch />
            </FlexItem>

            <FlexItem>
              <IstioStatus location={MASTHEAD} />
            </FlexItem>

            <FlexItem className={messageCenterStyle}>
              <MessageCenterTrigger />
            </FlexItem>

            <FlexItem className={helpDropdownStyle}>
              <HelpDropdown />
            </FlexItem>

            {serverConfig.kialiFeatureFlags.uiDefaults?.i18n?.showSelector && (
              <FlexItem className={languageSwitchStyle}>
                <LanguageSwitch />
              </FlexItem>
            )}

            <FlexItem data-test="user-dropdown" className={userDropdownStyle}>
              <UserDropdown />
            </FlexItem>
          </Flex>
        </ToolbarItem>
      </Toolbar>
    </>
  );
};
