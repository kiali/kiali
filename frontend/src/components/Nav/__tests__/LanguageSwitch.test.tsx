import * as React from 'react';
import { shallowToJson } from 'enzyme-to-json';
import { mount, shallow } from 'enzyme';
import { LanguageSwitchComponent } from '../Masthead/LanguageSwitch';
import { MenuToggle } from '@patternfly/react-core';
import { store } from 'store/ConfigStore';
import { Language } from 'types/Common';
import { serverConfig, setServerConfig } from 'config/ServerConfig';

const i18nServerConfig = Object.assign({}, serverConfig);

const delay = (ms: number): Promise<void> => new Promise(res => setTimeout(res, ms));

describe('Language switch', () => {
  beforeAll(() => {
    setServerConfig(i18nServerConfig);
  });
  it('renders correctly', () => {
    const wrapper = shallow(<LanguageSwitchComponent language={Language.ENGLISH} />);

    expect(shallowToJson(wrapper)).toBeDefined();
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('changes to english language', async () => {
    const wrapper = mount(<LanguageSwitchComponent language={Language.CHINESE} />);

    // click menu toggle
    wrapper.find(MenuToggle).simulate('click');

    // select English option
    wrapper
      .findWhere(node => node.key() === 'english')
      .findWhere(node => node.type() === 'button')
      .simulate('click');

    // wait a few ms for the language to be modified
    await delay(100);

    expect(store.getState().globalState.language).toBe(Language.ENGLISH);
  });

  it('changes to spanish language', async () => {
    const wrapper = mount(<LanguageSwitchComponent language={Language.ENGLISH} />);

    // click menu toggle
    wrapper.find(MenuToggle).simulate('click');

    // select Chinese option
    wrapper
      .findWhere(node => node.key() === 'spanish')
      .findWhere(node => node.type() === 'button')
      .simulate('click');

    // wait a few ms for the language to be modified
    await delay(100);

    expect(store.getState().globalState.language).toBe(Language.SPANISH);
  });

  it('changes to chinese language', async () => {
    const wrapper = mount(<LanguageSwitchComponent language={Language.ENGLISH} />);

    // click menu toggle
    wrapper.find(MenuToggle).simulate('click');

    // select Chinese option
    wrapper
      .findWhere(node => node.key() === 'chinese')
      .findWhere(node => node.type() === 'button')
      .simulate('click');

    // wait a few ms for the language to be modified
    await delay(100);

    expect(store.getState().globalState.language).toBe(Language.CHINESE);
  });
});
