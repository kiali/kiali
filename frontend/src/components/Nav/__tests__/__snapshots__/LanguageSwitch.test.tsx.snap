// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Language switch renders correctly 1`] = `
<Tooltip
  content={
    <React.Fragment>
      Switch language
    </React.Fragment>
  }
  exitDelay={0}
  position="bottom"
  trigger="mouseenter click"
>
  <Dropdown
    isOpen={false}
    onOpenChange={[Function]}
    onSelect={[Function]}
    popperProps={
      Object {
        "position": "center",
      }
    }
    toggle={[Function]}
  >
    <DropdownList>
      <DropdownItem
        key="english"
        onClick={[Function]}
      >
        <span>
          English
        </span>
        <Check
          className="kiali_fm5q8n4"
        />
      </DropdownItem>
      <DropdownItem
        key="spanish"
        onClick={[Function]}
      >
        <span>
          Español
        </span>
      </DropdownItem>
      <DropdownItem
        key="chinese"
        onClick={[Function]}
      >
        <span>
          中文
        </span>
      </DropdownItem>
    </DropdownList>
  </Dropdown>
</Tooltip>
`;
